kgraphql-ktor-stitched/src/main/kotlin/com/apurebase/kgraphql/stitched/schema/execution/AbstractRemoteRequestExecutor.kt

```diff
@@ -0,0 +1,172 @@
+package com.apurebase.kgraphql.stitched.schema.execution
+
+import com.apurebase.kgraphql.Context
+import com.apurebase.kgraphql.ExperimentalAPI
+import com.apurebase.kgraphql.GraphqlRequest
+import com.apurebase.kgraphql.request.Variables
+import com.apurebase.kgraphql.schema.execution.Execution
+import com.apurebase.kgraphql.schema.model.ast.OperationTypeNode
+import com.apurebase.kgraphql.schema.model.ast.SelectionNode
+import com.apurebase.kgraphql.schema.model.ast.TypeNode
+import com.apurebase.kgraphql.schema.model.ast.ValueNode
+import com.apurebase.kgraphql.schema.structure.Field
+import com.apurebase.kgraphql.schema.structure.Type
+import com.apurebase.kgraphql.stitched.RemoteExecutionException
+import com.apurebase.kgraphql.stitched.StitchedGraphqlRequest
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+
+@ExperimentalAPI
+abstract class AbstractRemoteRequestExecutor(private val objectMapper: ObjectMapper) : RemoteRequestExecutor {
+
+    /**
+     * Executes the actual [request] against the given [url] in the current [ctx]. This function is intended to
+     * be implemented by consumers to execute the actual request.
+     */
+    abstract suspend fun executeRequest(url: String, request: StitchedGraphqlRequest, ctx: Context): String
+
+    /**
+     * Main entry point called from the local request executor for the given [node] and [ctx].
+     */
+    final override suspend fun execute(node: Execution.Remote, ctx: Context): JsonNode? {
+        val remoteUrl = node.remoteUrl
+        val request = toGraphQLRequest(node, ctx)
+        val response = runCatching {
+            executeRequest(remoteUrl, request, ctx)
+        }.getOrElse {
+            """
+            { "errors": [ { "message": "${it.message}" } ] }
+            """.trimIndent()
+        }
+        val responseJson = objectMapper.readTree(response)
+        responseJson["errors"]?.let { errors ->
+            // TODO: properly transfer errors from the remote execution
+            val messages = (errors as? ArrayNode)?.map { (it as? ObjectNode)?.get("message")?.textValue() }
+                ?: listOf("Error(s) during remote execution")
+            throw RemoteExecutionException(message = messages.joinToString(", "), node = node)
+        }
+        return responseJson["data"]?.get(node.remoteOperation)
+    }
+
+    private fun SelectionNode.FieldNode.alias() = alias?.let {
+        "${it.value}: "
+    } ?: ""
+
+    private fun TypeNode.typeReference(): String {
```

Comment:
I wonder if this is generally true for `TypeNode` and could be an internal scoped extension function used anywhere in the code-base.

Reply:
It's not specific but it's also not used anywhere else so far. But I guess it makes sense to define it directly in `TypeNode`.